net: lwm2m: Improve thread safety of the library

rlubos opened the pull request:
With preemptive network threads enabled, the LwM2M could crash when for example LwM2M send was issued from another thread. This PR addresses the potential thread-safety issues detected with the library.

Fixes #78989

rlubos committed:
net: lwm2m: Make LwM2M message allocation thread safe

The LwM2M message allocation was not thread safe, i. e. the message was
acquired by setting the ctx pointer, and it was freed by clearing the
entire message structure.

If preemptive threads were enabled, and memset() clearing the message
content was preempted, the message structure being currently zeroed
migh've been allocated and initialized by some other thread. If the
thread releasing the message resumed work, it would continue clearing
the freshly allocated and initialized message structure.

In order to prevent this, introduce a new global mutex for the lwm2m
engine global variables. The mutex is used when LwM2M message is
allocated/deallocated. This will prevent reallocating the message
during the release process.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Make LwM2M context lists access thread safe

System lists are not thread safe, therefore all accesses should be
protected with a mutex. Introduce a LwM2M context specific mutex,
and use it whenever lists defined per-context are acessed.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

pdgendt commented on subsys/net/lib/lwm2m/lwm2m_engine.h:
```diff
@@ -375,4 +375,19 @@ int lwm2m_sock_nfds(void);
+ * @param[in] client_ctx LwM2M context
+ */
+void lwm2m_client_unlock(struct lwm2m_ctx *ctx);
+
```
Nit: some extraneous lines

rlubos replied:
Thanks, fixed

SeppoTakalo reviewed (dismissed):
This is acceptable, but I believe this is not all to make LwM2M thread safe.

We have also shared arrays in context
```
	struct coap_pending pendings[];
	struct coap_reply replies[];
```

Then `lwm2m_engine.c` have some global arrays that might be modified by API call from external thread
```
static struct service_node service_node_data[];
static sys_slist_t engine_service_list;
static struct zsock_pollfd sock_fds[];
static struct lwm2m_ctx *sock_ctx[];
static int sock_nfds;
static struct coap_block_context output_block_contexts[]
```

And `lwm2m_message_handling.c` also has some global arrays for messages
```
static struct lwm2m_message messages[];
static struct lwm2m_block_context block1_contexts[];
```

I did not do any analyze, if those actually are in danger, just a thought that we might need one global mutex to protect the engine as well. (And soon we have a danger of deadblock)

rlubos commented:
DNM for now, I'll update the PR according to suggestions from Seppo and retest.

rlubos committed:
net: lwm2m: Make CoAP pendings/replies arrays access thread safe

Make sure LwM2M context mutex is locked whenever accessing CoAP
pendings/replies arrays.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Make CoAP block contetx allocation thread safe

Use global engine mutex to protect allocation/deallocation of the CoAP
block contexts.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

pdgendt commented on subsys/net/lib/lwm2m/lwm2m_message_handling.c:
```diff
@@ -517,18 +540,25 @@ struct lwm2m_message *find_msg(struct coap_pending *pending, struct coap_reply *
 	return NULL;
 }
 
+
```
Nit: extra newline

rlubos replied:
Thanks again, fixed
